python/feathub/metric_stores/metric.py

```diff
@@ -0,0 +1,210 @@
+        """
+        self.metric_type = metric_type
+        self.window_size = window_size
+
+    def get_tags(self) -> Dict:
+        """
+        Gets the tags of this metric for tag-based metric reporters.
+        """
+        return {
+            "window_time": str(self.window_size),
+        }
+
+    @abstractmethod
+    def get_agg_transform(
+        self, feature_name: str, step_size: timedelta
+    ) -> Union[str, Transformation]:
+        """
+        Gets the transformation to aggregate and compute the metric value.
+
+        :param feature_name: The name of the host feature.
+        :param step_size: The time interval to get the aggregation result.
+        """
+        pass
+
+    @abstractmethod
+    def get_post_agg_transform(
+        self, agg_feature_name: str
+    ) -> Union[str, Transformation]:
+        """
+        Gets the transformation to post-process aggregation result.
+
+        :param agg_feature_name: The name of the feature containing aggregation result.
+        """
+        pass
+
+    @abstractmethod
+    def to_json(self) -> Dict:
+        """
+        Returns a json-formatted object representing this metric.
+        """
+        pass
+
+
+class Count(Metric):
+    """
+    A Metric that shows the number of features.
+    """
+
+    def __init__(
+        self,
+        filter_expr: Optional[str] = None,
+        window_size: timedelta = timedelta(seconds=0),
+    ):
+        """
+        :param filter_expr: Optional. If it is not None, it represents a partial FeatHub
+                            expression which evaluates to a boolean value. The partial
+                            Feathub expression should be a binary operator whose left
+                            child is absent and would be filled in with the host feature
+                            name. For example, "IS NULL" will be enriched into
+                            "{feature_name} IS NULL". Only features that evaluate this
+                            expression into True will be considered when computing the
+                            metric.
+        :param window_size: The time range to compute the metric. It should be zero or
+                            a positive time span. If it is zero, the metric will be
+                            computed from all features that have been processed since
+                            the Feathub job is created.
+        """
+        super().__init__("count", window_size)
+        self.filter_expr = filter_expr
+
+    def get_tags(self) -> Dict:
+        return {
+            **super(Count, self).get_tags(),
+            "filter_expr": self.filter_expr,
+        }
+
+    def get_agg_transform(
+        self, feature_name: str, step_size: timedelta
+    ) -> Union[str, Transformation]:
+        return SlidingWindowTransform(
+            expr="1",
+            agg_func="COUNT",
+            window_size=self.window_size,
+            step_size=step_size,
```

Comment:
Suppose window_size = 5 minutes and report_interval = 1 sec. This approach is going to use a sliding window with total of `5 min / 1 sec` = 300 steps. And the sliding window will need to keep track of up to 300 values internally.

Previously we discussed another approach where we can use one sliding window with window_size = step_size = 5 minute. And we can attach an extra sliding window with window_size = step_size = 1 sec AND agg_func = last. Then we will only need to maintain 2 values internally for this metric.

Also note that in general we don't expect report_interval to affect how the metric value is calculated. This current implementation breaks this expectation.

Reply:
I have implemented the behavior of window size = step size = 5 min.

Further appending a next sliding window to support report interval caused the following exception in tests: 
```
java.lang.RuntimeException: Record has Long.MIN_VALUE timestamp (= no timestamp marker). Is the time characteristic set to 'ProcessingTime', or did you forget to call 'DataStream.assignTimestampsAndWatermarks(...)'?
```
It seems that the upstream sliding window has lost timestamp attribute somewhere, but I haven't found the bug in code. I have added a TODO for supporting report interval.

Reply:
If it is not quick to fix, it might be simpler to add a UDF simply cache the last input value and periodically emits the value using processing time timer.

Reply:
`PythonUDFTransform` only supports simple mapping functions without states or timer service, so it might not be simple to achieve the UDF described above with existing infrastructure.

Reply:
Is it feasible to use Java UDF?